core/src/main/java/ai/timefold/solver/core/impl/util/IdentityElementAwareList.java

```diff
@@ -0,0 +1,93 @@
+package ai.timefold.solver.core.impl.util;
+
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.function.Consumer;
+import java.util.function.UnaryOperator;
+
+/**
+ * This data structure differs from {@link ElementAwareList} because it provides a find operation in O(H).
+ * 
+ * The implementation relies on two different data structures: {@link java.util.IdentityHashMap} and {@link ElementAwareList}.
+ *
+ * The element comparison works in the same way as {@link java.util.IdentityHashMap},
+ * and two keys k1 and k2 are considered equal if and only if (k1==k2).
+ *
+ * @param <T> The element type.
+ */
+public class IdentityElementAwareList<T> implements Iterable<T> {
```

Comment:
I've renamed the class to `IndexedElementAwareList`, and I've been thinking about that. The new list class will offer constant time for find, add, and remove operations.

Everything would work fine with a `List`, except for the operation `ExternalizedNextElementVariableSupply#getNext`, which will be O(N), and the new implementation is O(H).

I suggest maintaining this class as there will be performance benefits.

Reply:
Doesn't a `Map` also give a constant time? I'm not sure why we're inventing a new type here.

Reply:
Lukas, after double-checking the class, I believe it makes complete sense to maintain the index, previous, and next elements in the 'ExternalizedNextElementVariableSupply'.

If `LocationInList` also includes the previous and next elements, the class `IndexedElementAwareList` will not be necessary.

Reply:
Looking at the `ExternalizedNextElementVariableSupply` logic, I realized that we have implemented a similar feature, but I've used the `ElementAwareListEntry` instead of `LocationInList` and delegated the responsibility to a separate data structure class.

I wonder if we should move this logic to `ExternalizedListVariableStateSupply`. Otherwise, I will create a new "LocationInList" class with the next element reference and use a similar approach as `ExternalizedListVariableStateSupply`. What do you think?

Reply:
That's part of the reason why I suggested that `ExternalizedNextElementVariableSupply` and `ExternalizedListVariableStateSupply` shouldn't be separate. :-) It's the same logic, it should be in the same place.